File: 33. Search in Rotated Sorted Array.exs
#ELIXIR
#* 33. Search in Rotated Sorted Array
nums = [4,5,6,7,0,1,2]; target = 0
#SOL 1 | RUBY TRANSLATION | SUPER DUPER CONCISE | 100% | 266MS
defmodule Solution do
  @spec search(nums :: [integer], target :: integer) :: integer
  def search(nums, target) do
    Enum.find_index(nums,&(&1 == target)) || -1
  end
end

File: 33. Search in Rotated Sorted Array.rb
#* 33. Search in Rotated Sorted Array
nums = [4,5,6,7,0,1,2]; target = 0
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
#*
def search(nums, target)
  result = nums.sort.bsearch{ |n| target <=> n }
  result.nil? ? -1 : nums.index(result)
end
p search(nums, target)

#* LOL TO REMEMBER HOW To USE <=> UNDER HOOD
def search(nums, target)
  spaceship_operator_head_basher = ->(target, n) do
    return 0 if target == n
    return -1 if target < n
    return 1 if target > n
  end
  result = nums.sort.bsearch { |n| spaceship_operator_head_basher[target, n] }
  result.nil? ? -1 : nums.index(result)
end
p search(nums, target)